### DISCUSSION TITLE
remote workflow and tree_index

### DISCUSSION THREAD

**Discussion by JulesVandenbroeck** [columnflow/columnflow#383] [Q&A] [answered]
Hi everyone,

I have been building some custom tasks for fake-rate measurements. Since the event selection is quite soft the number of events passing is significant and merging these events using `cf.MergeReducedEvents` is not feasible (I could change the number of files merged but this results in messing with cf source code). Instead I am using the output of  `cf.ReduceEvents` to make histograms. I stumbled upon a problem when running this remote (htcondor): 'Exception: merge tree x not found, forest only contains 5 tree(s)'. It made me realise I do not know what the meaning/goal of tree_index is, is there any documentation on custom tasks and it's reqs or could someone clarify this variable?

The requirements of my custom task are written as:
```
def workflow_requires(self):
        reqs = super().workflow_requires()
        reqs["events"] = self.reqs.ReduceEvents.req(self)
        reqs["selection_stats"] = self.reqs.MergeSelectionStats.req(self, tree_index = 0, branch=-1)
        return reqs

    def requires(self):
        reqs = {
            "events": self.reqs.ReduceEvents.req(self),
            "selection_stats": self.reqs.MergeSelectionStats.req(self,tree_index = 0 ,branch=-1)
        }
```

**Comment by riga** [comment, marked as answer]
Hi @JulesVandenbroeck ,

we never came across this specific scenario, since we always had sufficiently low selection efficiencies so that merging was required, but I think your use case is perfectly valid and should be possible to skip using the merging step.

The merging happens in parallel, and if it is configured to combine 1000 files into 10, then for each of the 10 output files, 100 files are merged into 1. The merging logic itself comes with a law contrib tasks and is fully generic. Each output is the result of a merging tree (some more complex things happen in the background to merge on remote resources), and 10 files would lead to merging "trees" 0 - 9 (inclusive), with each of them being a workflow on its own.

The selection stats are always merged into 1 file (since it will be just a single json file in the end), so the `tree_index` is fixed to 0 here. The merging itself can be a workflow (possible running remotely), which is why `branch=-1`.

There will soon be a more in-depth documentation in law directly, so please exclude the current level of sparsity on that front.

Hope that helps!

**Reply by JulesVandenbroeck** [reply]
Hi Marcel,

Thank you for the response! It gives me a better understanding of what this `tree_index` means. In the mean time I found that my high selection efficiency was due to a bug rather than an extremely loose selection meaning I can go back to Merging the ReduceEvents outputs. The bug itself is a rather weird one where my selection is not the most recent version when ran remotely, with the warning:

fatal: ambiguous argument 'HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

**Reply by riga** [reply]
Oh, this sounds like an issue with your repository (I think), that is, as a result, not bundled correctly when sent with the jobs.
